File: scripts/asset-bench.py
#!/usr/bin/env python3
# L4 语料基准：1000 页 × L3 主题 + 图片资产（ImageSharp 解码/重编码）+ Sass 编译
# 目的：补已知局限 #2（图片管线从未进入测试）——Flint 自身资源管线画像
# 用法：python scripts/asset-bench.py [--pages 1000] [--images 100] [--runs 3]
import argparse, os, shutil, statistics, struct, subprocess, sys, time, zlib

BASE = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
FLINT = os.path.join(BASE, "benchmarks", "tools", "flint-aot", "Flint.exe")
ROOT = os.path.join(BASE, "benchmarks", "corpus", "asset-bench")

def make_png(path, w, h, seed):
    """手写噪声 PNG（zlib/struct，无依赖）：随机像素压不动，解码负载真实"""
    rng = seed
    raw = bytearray()
    for _ in range(h):
        raw.append(0)  # filter: none
        row = bytearray(w * 3)
        for x in range(w):
            rng = (1103515245 * rng + 12345) & 0x7FFFFFFF
            row[x*3] = rng & 0xFF
            row[x*3+1] = (rng >> 8) & 0xFF
            row[x*3+2] = (rng >> 16) & 0xFF
        raw.extend(row)
    def chunk(tag, data):
        c = struct.pack(">I", len(data)) + tag + data
        return c + struct.pack(">I", zlib.crc32(tag + data) & 0xFFFFFFFF)
    ihdr = struct.pack(">IIBBBBB", w, h, 8, 2, 0, 0, 0)
    with open(path, "wb") as f:
        f.write(b"\x89PNG\r\n\x1a\n")
        f.write(chunk(b"IHDR", ihdr))
        f.write(chunk(b"IDAT", zlib.compress(bytes(raw), 1)))
        f.write(chunk(b"IEND", b""))

SCSS = """$primary: #2b6cb0;
$radius: 4px;
@mixin card($pad: 12px) {
  padding: $pad;
  border-radius: $radius;
  box-shadow: 0 1px 3px rgba(0,0,0,.12);
}
@each $name, $color in (a: #e53e3e, b: #38a169, c: #d69e2e, d: #805ad5, e: #319795) {
  .badge-#{$name} { color: $color; border: 1px solid darken($color, 10%); @include card(6px); }
}
nav.site {
  ul { display: flex; gap: 8px; li { a { color: $primary; &:hover { text-decoration: underline; } } } }
}
@for $i from 1 through 40 {
  .col-#{$i} { width: 100% / $i; @if $i % 5 == 0 { border-left: 1px solid #eee; } }
}
article { @include card; h1 { font-size: 1.6rem; color: $primary; } p code { background: #f7fafc; } }
"""

def gen_corpus(site, pages, images):
    content = os.path.join(site, "content", "posts")
    layouts = os.path.join(site, "layouts", "_default")
    partials = os.path.join(site, "layouts", "partials")
    imgdir = os.path.join(site, "assets", "images")
    styledir = os.path.join(site, "assets", "styles")
    for d in (content, layouts, partials, imgdir, styledir):
        os.makedirs(d, exist_ok=True)
    with open(os.path.join(site, "Flint.toml"), "w", encoding="utf-8") as f:
        f.write('baseURL = "http://localhost:1313/"\ntitle = "Asset Bench"\n')
    with open(os.path.join(layouts, "single.html"), "w", encoding="utf-8") as f:
        f.write('<nav>{{ include "nav" }}</nav>{{ include "sidebar" }}<article><h1>{{ page.title }}</h1>{{ page.content }}</article>{{ include "related" }}{{ partialcached "footer" }}')
    with open(os.path.join(partials, "sidebar.html"), "w", encoding="utf-8") as f:
        f.write('<aside>{{ for p in site.regular_pages }}<li><a href="{{ p.rel_permalink }}">{{ p.title }}</a></li>{{ end }}</aside>')
    with open(os.path.join(partials, "nav.html"), "w", encoding="utf-8") as f:
        f.write('<nav>{{ for p in site.regular_pages }}<a href="{{ p.rel_permalink }}">{{ p.title }}</a>{{ end }}</nav>')
    with open(os.path.join(partials, "related.html"), "w", encoding="utf-8") as f:
        f.write('<div>{{ for p in site.regular_pages }}{{ if p.type == "page" }}<span>{{ p.title }}</span>{{ end }}{{ end }}</div>')
    with open(os.path.join(partials, "footer.html"), "w", encoding="utf-8") as f:
        f.write("<footer>C</footer>")
    with open(os.path.join(site, "layouts", "index.html"), "w", encoding="utf-8") as f:
        f.write("<h1>home</h1>")
    for i in range(pages):
        with open(os.path.join(content, f"p{i:04d}.md"), "w", encoding="utf-8") as f:
            f.write(f'---\ntitle: "Post {i}"\ndate: 2026-01-01\ntags: ["t{i % 20}"]\n---\n\nBody {i} lorem ipsum.')
    print(f"生成 {images} 张 PNG（640x480 噪声）...", file=sys.stderr)
    for i in range(images):
        make_png(os.path.join(imgdir, f"img_{i:03d}.png"), 640, 480, seed=i + 7)
    with open(os.path.join(styledir, "main.scss"), "w", encoding="utf-8") as f:
        f.write(SCSS)

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--pages", type=int, default=1000)
    ap.add_argument("--images", type=int, default=100)
    ap.add_argument("--runs", type=int, default=3)
    args = ap.parse_args()
    site = os.path.join(ROOT, "site")
    pub = os.path.join(ROOT, "site-pub")
    if os.path.exists(site):
        shutil.rmtree(site)
    gen_corpus(site, args.pages, args.images)
    print(f"L4 语料就绪：{args.pages} 页 × L3 主题 + {args.images} 图 + Sass", file=sys.stderr)
    import psutil
    times, peaks = [], []
    for i in range(args.runs + 1):  # 首轮预热（含磁盘缓存建立）
        shutil.rmtree(pub, ignore_errors=True)
        t0 = time.perf_counter()
        env = {**os.environ, "FLINT_TRACE_PHASES": "1"}
        p = subprocess.Popen([FLINT, "build", "-s", site, "-o", pub],
                             stdout=subprocess.DEVNULL, stderr=subprocess.PIPE, env=env)
        proc = psutil.Process(p.pid)
        peak = 0
        while p.poll() is None:
            try:
                peak = max(peak, proc.memory_info().rss)
            except (psutil.NoSuchProcess, psutil.AccessDenied):
                pass
            time.sleep(0.02)
        el = (time.perf_counter() - t0) * 1000
        assert p.returncode == 0, f"构建失败 rc={p.returncode}"
        peaks.append(peak / 2**20)
        if i > 0:
            times.append(el)
        if i == 0:
            print("--- 预热轮阶段明细 ---", file=sys.stderr)
            print(p.stderr.read().decode("utf-8", "replace"), file=sys.stderr)
    print(f"\n=== L4 资源管线画像（{args.runs} 轮中位）===")
    print(f"构建时间中位: {statistics.median(times):.0f}ms")
    print(f"内存峰值中位: {statistics.median(peaks):.0f}MB")
    html = sum(len(files) for _, _, files in os.walk(pub))
    print(f"产出文件数: {html}")

if __name__ == "__main__":
    main()

File: src/Flint.DevTools/Bench/AssetBench.cs
// bench asset：L4 资源管线画像（scripts/asset-bench.py 的 C# 移植）
// 1000 页 × L3 主题 + 图片资产（噪声 PNG 解码）+ Sass 编译；
// 手写 LCG 噪声 PNG（与 Python 版同种子同像素公式；压缩流由 .NET zlib 产生，
// 文件字节与 Python 版可不同但像素内容一致，对解码负载等价）。

using System.Buffers.Binary;
using System.CommandLine;
using System.IO.Compression;

namespace Flint.DevTools.Bench;

/// <summary>
/// 资源管线基准
/// </summary>
internal static class AssetBench
{
    private const string Scss = @"$primary: #2b6cb0;
$radius: 4px;
@mixin card($pad: 12px) {
  padding: $pad;
  border-radius: $radius;
  box-shadow: 0 1px 3px rgba(0,0,0,.12);
}
@each $name, $color in (a: #e53e3e, b: #38a169, c: #d69e2e, d: #805ad5, e: #319795) {
  .badge-#{$name} { color: $color; border: 1px solid darken($color, 10%); @include card(6px); }
}
nav.site {
  ul { display: flex; gap: 8px; li { a { color: $primary; &:hover { text-decoration: underline; } } } }
}
@for $i from 1 through 40 {
  .col-#{$i} { width: 100% / $i; @if $i % 5 == 0 { border-left: 1px solid #eee; } }
}
article { @include card; h1 { font-size: 1.6rem; color: $primary; } p code { background: #f7fafc; } }
";

    internal static Command BuildCommand()
    {
        var pagesOpt = new Option<int>("--pages") { Description = "页数", DefaultValueFactory = _ => 1000 };
        var imagesOpt = new Option<int>("--images") { Description = "图片数（640x480 噪声 PNG）", DefaultValueFactory = _ => 100 };
        var runsOpt = new Option<int>("--runs") { Description = "测量轮数（不含预热）", DefaultValueFactory = _ => 3 };

        var cmd = new Command("asset", "L4 资源管线画像（图片解码/重编码 + Sass 编译）");
        cmd.Options.Add(pagesOpt);
        cmd.Options.Add(imagesOpt);
        cmd.Options.Add(runsOpt);

        cmd.SetAction(async parseResult =>
        {
            var pages = parseResult.GetValue(pagesOpt);
            var images = parseResult.GetValue(imagesOpt);
            var runs = parseResult.GetValue(runsOpt);

            var root = Path.GetFullPath(DefaultPaths.AssetBenchRoot);
            var site = Path.Combine(root, "site");
            var pub = Path.Combine(root, "site-pub");

            RepoGuard.DeleteTree(site);
            GenCorpus(site, pages, images);
            Console.Error.WriteLine($"L4 语料就绪：{pages} 页 × L3 主题 + {images} 图 + Sass");

            ProcessRunner.RequireExe(DefaultPaths.FlintExe);

            var times = new List<double>(runs);
            var peaks = new List<double>(runs);
            var flintArgs = new[] { "build", "-s", site, "-o", pub };
            var env = new Dictionary<string, string> { ["FLINT_TRACE_PHASES"] = "1" };

            for (var i = 0; i < runs + 1; i++) // 首轮预热（含磁盘缓存建立）
            {
                RepoGuard.DeleteTree(pub);
                var result = await ProcessRunner.RunTimedAsync(DefaultPaths.FlintExe, flintArgs, env).ConfigureAwait(false);
                if (result.ExitCode != 0)
                {
                    throw new InvalidOperationException($"构建失败 rc={result.ExitCode}: {result.Stderr}");
                }

                peaks.Add(result.PeakRssBytes / 1048576.0);

                if (i == 0)
                {
                    Console.Error.WriteLine("--- 预热轮阶段明细 ---");
                    Console.Error.WriteLine(result.Stderr);
                }
                else
                {
                    times.Add(result.ElapsedMs);
                }
            }

            Console.WriteLine();
            Console.WriteLine($"=== L4 资源管线画像（{runs} 轮中位）===");
            Console.WriteLine($"构建时间中位: {ProcessRunner.Median(times):F0}ms");
            Console.WriteLine($"内存峰值中位: {ProcessRunner.Median(peaks):F0}MB");
            Console.WriteLine($"产出文件数: {CountFiles(pub)}");
            return 0;
        });

        return cmd;
    }

    private static int CountFiles(string dir)
    {
        return Directory.Exists(dir)
            ? Directory.EnumerateFiles(dir, "*", SearchOption.AllDirectories).Count()
            : 0;
    }

    private static void GenCorpus(string site, int pages, int images)
    {
        var content = Path.Combine(site, "content", "posts");
        var layouts = Path.Combine(site, "layouts", "_default");
        var partials = Path.Combine(site, "layouts", "partials");
        var imgdir = Path.Combine(site, "assets", "images");
        var styledir = Path.Combine(site, "assets", "styles");
        foreach (var d in new[] { content, layouts, partials, imgdir, styledir })
        {
            Directory.CreateDirectory(d);
        }

        File.WriteAllText(Path.Combine(site, "Flint.toml"),
            "baseURL = \"http://localhost:1313/\"\ntitle = \"Asset Bench\"\n", Utf8.NoBom);
        File.WriteAllText(Path.Combine(layouts, "single.html"),
            "<nav>{{ include \"nav\" }}</nav>{{ include \"sidebar\" }}<article><h1>{{ page.title }}</h1>{{ page.content }}</article>{{ include \"related\" }}{{ partialcached \"footer\" }}",
            Utf8.NoBom);
        File.WriteAllText(Path.Combine(partials, "sidebar.html"),
            "<aside>{{ for p in site.regular_pages }}<li><a href=\"{{ p.rel_permalink }}\">{{ p.title }}</a></li>{{ end }}</aside>", Utf8.NoBom);
        File.WriteAllText(Path.Combine(partials, "nav.html"),
            "<nav>{{ for p in site.regular_pages }}<a href=\"{{ p.rel_permalink }}\">{{ p.title }}</a>{{ end }}</nav>", Utf8.NoBom);
        File.WriteAllText(Path.Combine(partials, "related.html"),
            "<div>{{ for p in site.regular_pages }}{{ if p.type == \"page\" }}<span>{{ p.title }}</span>{{ end }}{{ end }}</div>", Utf8.NoBom);
        File.WriteAllText(Path.Combine(partials, "footer.html"), "<footer>C</footer>", Utf8.NoBom);
        File.WriteAllText(Path.Combine(site, "layouts", "index.html"), "<h1>home</h1>", Utf8.NoBom);

        for (var i = 0; i < pages; i++)
        {
            var fm = $"---\ntitle: \"Post {i}\"\ndate: 2026-01-01\ntags: [\"t{i % 20}\"]\n---\n\nBody {i} lorem ipsum.";
            File.WriteAllText(Path.Combine(content, $"p{i:D4}.md"), fm, Utf8.NoBom);
        }

        Console.Error.WriteLine($"生成 {images} 张 PNG（640x480 噪声）...");
        for (var i = 0; i < images; i++)
        {
            NoisePng.Write(Path.Combine(imgdir, $"img_{i:D3}.png"), 640, 480, i + 7);
        }

        File.WriteAllText(Path.Combine(styledir, "main.scss"), Scss, Utf8.NoBom);
    }

    /// <summary>手写噪声 PNG：LCG 像素（与 Python 版同种子同公式，像素内容逐字节一致）；
    /// zlib 压缩流由 .NET 实现产生，文件字节与 Python 版可不同，对解码负载等价。</summary>
    private static class NoisePng
    {
        private static readonly byte[] PngSignature = { 0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A };
        private static readonly uint[] CrcTable = BuildCrcTable();

        public static void Write(string path, int w, int h, int seed)
        {
            var rng = (uint)seed;
            var raw = new byte[h * (1 + (w * 3))];
            var pos = 0;
            for (var y = 0; y < h; y++)
            {
                raw[pos++] = 0; // filter: none
                for (var x = 0; x < w; x++)
                {
                    rng = ((1103515245 * rng) + 12345) & 0x7FFFFFFF;
                    raw[pos++] = (byte)(rng & 0xFF);
                    raw[pos++] = (byte)((rng >> 8) & 0xFF);
                    raw[pos++] = (byte)((rng >> 16) & 0xFF);
                }
            }

            using var fs = File.Create(path);
            fs.Write(PngSignature);

            var ihdr = new byte[13];
            BinaryPrimitives.WriteUInt32BigEndian(ihdr.AsSpan(0, 4), (uint)w);
            BinaryPrimitives.WriteUInt32BigEndian(ihdr.AsSpan(4, 4), (uint)h);
            ihdr[8] = 8;  // bit depth
            ihdr[9] = 2;  // color type: truecolor
            ihdr[10] = 0; // compression
            ihdr[11] = 0; // filter
            ihdr[12] = 0; // interlace
            WriteChunk(fs, "IHDR", ihdr);

            using var idat = new MemoryStream();
            using (var zlib = new ZLibStream(idat, CompressionLevel.Fastest, leaveOpen: true))
            {
                zlib.Write(raw);
            }

            WriteChunk(fs, "IDAT", idat.ToArray());
            WriteChunk(fs, "IEND", Array.Empty<byte>());
        }

        private static void WriteChunk(Stream stream, string tag, byte[] data)
        {
            var tagBytes = System.Text.Encoding.ASCII.GetBytes(tag);

            uint crc = 0xFFFFFFFF;
            foreach (var b in tagBytes)
            {
                crc = (crc >> 8) ^ CrcTable[(crc ^ b) & 0xFF];
            }

            foreach (var b in data)
            {
                crc = (crc >> 8) ^ CrcTable[(crc ^ b) & 0xFF];
            }

            crc ^= 0xFFFFFFFF;

            Span<byte> header = stackalloc byte[4];
            BinaryPrimitives.WriteUInt32BigEndian(header, (uint)data.Length);
            stream.Write(header);
            stream.Write(tagBytes);
            stream.Write(data);

            Span<byte> crcSpan = stackalloc byte[4];
            BinaryPrimitives.WriteUInt32BigEndian(crcSpan, crc);
            stream.Write(crcSpan);
        }

        private static uint[] BuildCrcTable()
        {
            var table = new uint[256];
            for (uint n = 0; n < 256; n++)
            {
                var c = n;
                for (var k = 0; k < 8; k++)
                {
                    c = (c & 1) != 0 ? (0xEDB88320 ^ (c >> 1)) : (c >> 1);
                }

                table[n] = c;
            }

            return table;
        }
    }
}
